iceoryx_hoofs/filesystem/include/iox/filesystem.hpp

```diff
@@ -116,136 +117,62 @@ class access_rights final
-        }
-
-        /// @note Implementation detail! Please use 'iox::perms::all'.
-        static constexpr access_rights all() noexcept
-        {
-            constexpr value_type VALUE{0777};
-            return access_rights{VALUE};
-        }
-
-        /// @note Implementation detail! Please use 'iox::perms::set_uid'.
-        static constexpr access_rights set_uid() noexcept
-        {
-            constexpr value_type VALUE{04000};
-            return access_rights{VALUE};
-        }
-        /// @note Implementation detail! Please use 'iox::perms::set_gid'.
-        static constexpr access_rights set_gid() noexcept
-        {
-            constexpr value_type VALUE{02000};
-            return access_rights{VALUE};
-        }
-        /// @note Implementation detail! Please use 'iox::perms::sticky_bit'.
-        static constexpr access_rights sticky_bit() noexcept
-        {
-            constexpr value_type VALUE{01000};
-            return access_rights{VALUE};
-        }
-
-        /// @note Implementation detail! Please use 'iox::perms::mask'.
-        static constexpr access_rights mask() noexcept
-        {
-            constexpr value_type VALUE{07777};
-            return access_rights{VALUE};
-        }
-
-        /// @note Implementation detail! Please use 'iox::perms::unknown'.
-        static constexpr access_rights unknown() noexcept
-        {
-            constexpr value_type VALUE{0xFFFFU};
-            return access_rights{VALUE};
-        }
+        static constexpr value_type NONE{0};
+
+        static constexpr value_type OWNER_READ{0400};
+        static constexpr value_type OWNER_WRITE{0200};
+        static constexpr value_type OWNER_EXEC{0100};
+        static constexpr value_type OWNER_ALL{0700};
+
+        static constexpr value_type GROUP_READ{040};
+        static constexpr value_type GROUP_WRITE{020};
+        static constexpr value_type GROUP_EXEC{010};
+        static constexpr value_type GROUP_ALL{070};
+
+        static constexpr value_type OTHERS_READ{04};
+        static constexpr value_type OTHERS_WRITE{02};
+        static constexpr value_type OTHERS_EXEC{01};
+        static constexpr value_type OTHERS_ALL{07};
+
+        static constexpr value_type ALL{0777};
+
+        static constexpr value_type SET_UID{04000};
+        static constexpr value_type SET_GID{02000};
+        static constexpr value_type STICKY_BIT{01000};
+
+        static constexpr value_type MASK{07777};
+
+        // intentionally different from 'std::filesystem::perms::unknown' to prevent unexpected results. Combining a
+        // permission set to 'std::filesystem::perms::unknown' with other permission flags using bitwise OR may result
+        // in a value that is not representative of a valid permission state. By setting the value to '0x8000' only the
+        // MSB is 1 and all bits representing permissions are set to 0 and a bitwise OR will therefore also always
+        // result in a 0.
```

Comment:
Maybe I got this wrong, but with this change, the result of the bitwise OR actually depends on the other permission flag **and** is valid, right?

Reply:
Aaaarg, I meant bitwise AND :sweat_smile: 

Bitwise OR still behaves like before but bitwise AND with `perms::unknown` will evaluate to `false` when the rhs isn't a `perms::unknown`. Similarly bitwise XOR with `perms::unknown` will evaluate to `true` when the rhs isn't a `perms::unknown`.

Reply:
Aaaarg. I think I understand now why the STL chose `0xFFFF`. Once a bit is reset it isn't `unknown` anymore but with `0x8000` the MSB would still indicate the unknown state. Storing the result of `0x8000 | 0x0001` in a variable and using it later on will have an ambiguous bit pattern. I guess the STL did not want to decide on the outcome of such operations. With `0xFFFF` they circumvented the problem by defining the value of `0xFFFF` as `unknown` and `0xFFFE` would therefore not be considered unknown as long as bit 0 is not set to 1.

I think we could do similar things by also allowing values like `0x8001`. I have to think more about it to ensure there are no issues with this representation.

Do you see any other problems?

Reply:
I rethought the free functions a bit. There is no need for someone to create an `access_rights` object from a value with the `unknown` flag set. I also updated the documentation and I think `0x8000` is a more safe value than `0xFFFF` to indicate a unknown value since this would result in no rights when used in a wrong way.